lib/std/crypto/25519/edwards25519.zig

```diff
@@ -250,8 +252,138 @@ pub const Edwards25519 = struct {
         scalar.clamp(&t);
         return mul(p, t);
     }
+
+    // montgomery -- recover y = sqrt(x^3 + A*x^2 + x)
+    fn xmontToYmont(x: Fe) !Fe {
+        var x2 = x.sq();
+        const x3 = x.mul(x2);
+        x2 = x2.mul32(Fe.edwards25519a_32);
+        return x.add(x2).add(x3).sqrt();
+    }
+
+    // montgomery affine coordinates to edwards extended coordinates
+    fn montToEd(x: Fe, y: Fe) Edwards25519 {
+        const x_plus_one = x.add(Fe.one);
+        const x_minus_one = x.sub(Fe.one);
+        const x_plus_one_y_inv = x_plus_one.mul(y).invert(); // 1/((x+1)*y)
+
+        // xed = sqrt(-A-2)*x/y
+        const xed = x.mul(Fe.edwards25519sqrtam2).mul(x_plus_one_y_inv).mul(x_plus_one);
+
+        // yed = (x-1)/(x+1) or 1 if the denominator is 0
+        var yed = x_plus_one_y_inv.mul(y).mul(x_minus_one);
+        yed.cMov(Fe.one, @boolToInt(x_plus_one_y_inv.isZero()));
+
+        return Edwards25519{
+            .x = xed,
+            .y = yed,
+            .z = Fe.one,
+            .t = xed.mul(yed),
+        };
+    }
+
+    /// Elligator2 map - Returns Montgomery affine coordinates
+    pub fn elligator2(r: Fe) struct { x: Fe, y: Fe, not_square: bool } {
+        const rr2 = r.sq2().add(Fe.one).invert();
+        var x = rr2.mul32(Fe.edwards25519a_32).neg(); // x=x1
+        var x2 = x.sq();
+        const x3 = x2.mul(x);
+        x2 = x2.mul32(Fe.edwards25519a_32); // x2 = A*x1^2
+        const gx1 = x3.add(x).add(x2); // gx1 = x1^3 + A*x1^2 + x1
+        const not_square = !gx1.isSquare();
+
+        // gx1 not a square => x = -x1-A
+        x.cMov(x.neg(), @boolToInt(not_square));
+        x2 = Fe.zero;
+        x2.cMov(Fe.edwards25519a, @boolToInt(not_square));
+        x = x.sub(x2);
+
+        // We have y = sqrt(gx1) or sqrt(gx2) with gx2 = gx1*(A+x1)/(-x1)
+        // but it is about as fast to just recompute y from the curve equation.
+        const y = xmontToYmont(x) catch unreachable;
+        return .{ .x = x, .y = y, .not_square = not_square };
+    }
+
+    /// Map a 64-bit hash into an Edwards25519 point
+    pub fn fromHash(h: [64]u8) Edwards25519 {
```

Comment:
Is there an use for a 32-bit version?

Reply:
You can use `elligator2()` directly. That function returns affine coordinates, but a simple `toBytes()`/`fromBytes()` dance on the Y coordinate will get you an `Edwards25519` structure from it.

libsodium has a 32 bit version, that uses the elligator2 map from the 255 lower bits, and set the sign of X from the top bit. It is way faster than the H2C operation. Maybe a compatible function can be added to Zig, but now that a standard exists, applications should probably use it even if it's overkill for 25519.

Reply:
I added a compatible `fromUniform()` function, which is non-standard, but can be useful when you need a random a random point and don't want to go through the H2C machinery.

Reply:
It was just a comment out of curiosity. You are gonna have a hard time finding someone worthy of properly reviewing your PR.

Reply:
Thanks for your feedback!

I'll talk about the 25519 stuff, especially implementation details, on Saturday 22's Zig show.

Stay tuned!